### ISSUE TITLE
Cannot detect Kubernetes Secrets changes

### ISSUE THREAD

**Issue by snsinfu** [snsinfu/ftrap#3]
I want to use ftrap to detect Kubernetes Secrets changes in a container but it does not work. I used this CMD directive:

```
CMD ["ftrap", "-f", "/srv/secrets", "myapp"]
```

with my secrets mounted on /srv/secrets. ftrap did not send SIGHUP to myapp when I deleted or re-created the secrets.

I entered into the container and used fswatch to monitor inotify events. I got this:

```
Mon Sep  2 20:34:11 2019 /srv/secrets IsDir
Mon Sep  2 20:34:11 2019 /srv/secrets/..2019_09_02_20_31_28.101389353 PlatformSpecific
Mon Sep  2 20:34:12 2019 /srv/secrets IsDir
Mon Sep  2 20:34:12 2019 /srv/secrets/..2019_09_02_20_31_28.101389353 PlatformSpecific
Mon Sep  2 20:34:12 2019 /srv/secrets IsDir
Mon Sep  2 20:34:12 2019 /srv/secrets/..2019_09_02_20_31_28.101389353 PlatformSpecific
Mon Sep  2 20:34:12 2019 /srv/secrets IsDir
Mon Sep  2 20:34:12 2019 /srv/secrets/..2019_09_02_20_34_11.350778740 Created
Mon Sep  2 20:34:12 2019 /srv/secrets IsDir
Mon Sep  2 20:34:12 2019 /srv/secrets/..2019_09_02_20_34_11.350778740 AttributeModified
Mon Sep  2 20:34:12 2019 /srv/secrets/..data_tmp Created
Mon Sep  2 20:34:12 2019 /srv/secrets/..data_tmp Removed MovedFrom
Mon Sep  2 20:34:12 2019 /srv/secrets/..data Created MovedTo
Mon Sep  2 20:34:12 2019 /srv/secrets/config.json Removed
Mon Sep  2 20:34:12 2019 /srv/secrets IsDir
Mon Sep  2 20:34:12 2019 /srv/secrets PlatformSpecific
Mon Sep  2 20:34:12 2019 /srv/secrets IsDir
Mon Sep  2 20:34:12 2019 /srv/secrets/..2019_09_02_20_31_28.101389353 PlatformSpecific
Mon Sep  2 20:34:12 2019 /srv/secrets IsDir
Mon Sep  2 20:34:12 2019 /srv/secrets/..2019_09_02_20_31_28.101389353 PlatformSpecific
Mon Sep  2 20:34:12 2019 /srv/secrets IsDir
Mon Sep  2 20:34:12 2019 /srv/secrets/..2019_09_02_20_31_28.101389353 PlatformSpecific
Mon Sep  2 20:34:12 2019 /srv/secrets IsDir
Mon Sep  2 20:34:12 2019 /srv/secrets/..2019_09_02_20_31_28.101389353 Removed
Mon Sep  2 20:34:12 2019 /srv/secrets IsDir
Mon Sep  2 20:34:12 2019 /srv/secrets PlatformSpecific
Mon Sep  2 20:35:36 2019 /srv/secrets IsDir
Mon Sep  2 20:35:36 2019 /srv/secrets/..2019_09_02_20_34_11.350778740 PlatformSpecific
Mon Sep  2 20:35:37 2019 /srv/secrets IsDir
Mon Sep  2 20:35:37 2019 /srv/secrets/..2019_09_02_20_34_11.350778740 PlatformSpecific
Mon Sep  2 20:35:37 2019 /srv/secrets IsDir
Mon Sep  2 20:35:37 2019 /srv/secrets/..2019_09_02_20_34_11.350778740 PlatformSpecific
Mon Sep  2 20:35:37 2019 /srv/secrets IsDir
Mon Sep  2 20:35:37 2019 /srv/secrets/..2019_09_02_20_35_36.875582531 Created
Mon Sep  2 20:35:37 2019 /srv/secrets IsDir
Mon Sep  2 20:35:37 2019 /srv/secrets/..2019_09_02_20_35_36.875582531 AttributeModified
Mon Sep  2 20:35:37 2019 /srv/secrets/config.json Created
Mon Sep  2 20:35:37 2019 /srv/secrets/..data_tmp Created
Mon Sep  2 20:35:37 2019 /srv/secrets/..data_tmp Removed MovedFrom
Mon Sep  2 20:35:37 2019 /srv/secrets/..data Created MovedTo
Mon Sep  2 20:35:37 2019 /srv/secrets IsDir
Mon Sep  2 20:35:37 2019 /srv/secrets/..2019_09_02_20_34_11.350778740 Removed
```

Looks like no Modified event is fired. K8s just removes the old secret and creates new one.

**Comment by snsinfu** [comment]
So, probably ftrap needs to watch more events like `IN_CREATE`. I'm not sure what events to watch though ([list][1]). It should depend on whether the watched path is a file or a directory. For files `IN_CREATE | IN_DELETE | IN_MODIFY` would be interesting. For directories `IN_MOVED_FROM | IN_MOVED_TO` would also be interesting, but don't these MOVED events suggest DELETE or CREATE events? Is DELETE events really interesting?

[1]: https://linux.die.net/man/7/inotify

**Comment by snsinfu** [comment]
The main purpose of using ftrap is config file reloading, so DELETE event is not useful. Watching CREATE/MODIFY events should do.

Note: Do not add a command-line option to specify which events to watch. Keep ftrap command simple. One can always script with fswatch for ultimate flexibility.

**Comment by snsinfu** [comment]
It's way more complicated than I thought. See these SO threads:

- https://stackoverflow.com/questions/13409843/inotify-vim-modification
- https://stackoverflow.com/questions/18890116/inotify-does-not-raise-delete-self-if-an-open-file-fd-exists

MOVED_TO, DELETE_SELF and IGNORED events are delivered when a watched file is clobbered, and the watch descriptor is automatically removed from the inotify instance. So we need to re-add the file to the inotify instance. But the events would not be sent when the file has hardlink(s)...

And worse, as we see in the fswatch log above, Secrets-mounted data is first removed and then re-created (not clobbered) after a considerable amount of delay. So we can't immediately start rewatching the file.

Maybe we need to always watch the containing directory if the target is a file, and use CREATE, MODIFY and CLOSE_WRITE events on the file...

If the directory is renamed or deleted, we have nothing to do... maybe periodically monitor the target path and start watching new directory when a new one appeared?

**Comment by snsinfu** [comment]
We can use the timeout feature of poll(2). Have a list of target files with "is-watched" flags. Add as many files as possible to inotify. Poll on the inotify, and on timeout, try re-adding not-yet-watched target files to the inotify. Repeat this.

**Comment by snsinfu** [comment]
Investigated: https://github.com/snsinfu/bit4/tree/master/test862-inotify_remove_create
Several notes.

### When watching a file

- `DELETE_SELF` means the file is deleted. `IGNORED` event follows and the file is automatically unwatched.
- After `MOVE_SELF` the watched file does not exist at the original path. Treat it as a delete event and unwatch the file.
- As noted above `ATTRIB` can actually be a delete event when the watched file has hardlink(s). Check the existence of the original path.

### When watching a directory

- `CREATE`, `DELETE`, `MOVED_FROM` and `MOVED_TO` events are fired in addition to the file events.

### Implementation note

- `read()` may read multiple events from an inotify file descriptor. Loop and process all events in the buffer!
- Looks like watch descriptor is not reused after auto-unwatch (IGNNORED event). The number just grows.

**Comment by snsinfu** [comment]
Hacked: b24e1643a8a25fd3cc2ad99ec5942dade12487f6
Should be fixed in [v0.2.0](https://github.com/snsinfu/ftrap/releases/tag/v0.2.0).
